File: aoc2018/day9/solve.js
function solve(players, lastMarble) {
  const mkNode = value => ({
    prev: null,
    next: null,
    value
  });

  // We create a circular list of one node
  let root = mkNode(0);
  root.next = root;
  root.prev = root;

  let currentMarble = 1;
  let currentPlayer = 0;
  const scores = new Uint32Array(players);

  // Start the game
  while (currentMarble <= lastMarble) {
    if (currentMarble % 23 === 0) {
      // Move 7 spots counter-clockwise
      for (let i = 0; i < 7; i += 1) {
        root = root.prev;
      }

      scores[currentPlayer] += currentMarble + root.value;

      // Remove current node
      root.prev.next = root.next;
      root.next.prev = root.prev;
      root = root.next;
    } else {
      // Move one step clockwise
      root = root.next;

      // Insert node
      const newNode = mkNode(currentMarble);
      newNode.next = root.next;
      newNode.prev = root;
      root.next.prev = newNode;
      root.next = newNode;

      root = newNode;
    }

    currentMarble += 1;
    currentPlayer = (currentPlayer + 1) % players;
  }

  return Math.max(...scores);
}

function assert(players, lastMarble, expected) {
  const t0 = Date.now();
  const result = solve(players, lastMarble);
  const total = Date.now() - t0;
  if (result !== expected) {
    console.error("Wrong", players, lastMarble, result, expected);
  } else {
    console.log("Test", {
      players,
      lastMarble,
      total
    });
  }
}

assert(9, 25, 32);
assert(10, 1618, 8317);
assert(13, 7999, 146373);
assert(17, 1104, 2764);
assert(21, 6111, 54718);
assert(30, 5807, 37305);

console.log("Part 1:", solve(435, 71184));
console.log("Part 2:", solve(435, 71184 * 100));

File: aoc2018/day9/solve.hs

import qualified Data.IntMap.Strict as M
import qualified Data.List as L

data LList = LList [Int] Int [Int]
--                 ^     ^   ^ right
--                 |     | focused element
--                 | left

-- Perform one clockwise rotation on the list
clockwise :: LList -> LList
clockwise llist@(LList [] _ []) = llist
clockwise (LList left v (x:xs)) = LList (v:left) x xs
clockwise (LList left v []) = let (x:xs) = reverse left in LList [] x (xs ++ [v])

-- Perform one counter-clockwise rotation on the list
counterClockwise :: LList -> LList
counterClockwise llist@(LList [] _ []) = llist
counterClockwise (LList (x:xs) v right) = LList xs x (v:right)
counterClockwise (LList [] v right) = let (x:xs) = reverse right in LList (xs ++ [v]) x []

-- Perform 7 counter-clockwise rotations on the list
counterClockwise7 :: LList -> LList
counterClockwise7 = (
    counterClockwise
    . counterClockwise
    . counterClockwise
    . counterClockwise
    . counterClockwise
    . counterClockwise
    . counterClockwise)

-- Insert an element on the right of the currently focused element
insert :: LList -> Int -> LList
insert (LList left v right) value = LList (v:left) value right

-- Remove the currently focused element
remove :: LList -> LList
remove (LList left _ (x:xs)) = LList left x xs
remove (LList left _ []) = let (x:xs) = reverse left in LList [] x xs

-- Return currently focused element
focus :: LList -> Int
focus (LList _ v _) = v

solve :: Int -> Int -> Int
solve players lastMarble = go root 1 0 (M.fromList [(p, 0) | p <- [0..players-1]])
    where
        -- Initial marble
        root = LList [] 0 []
        -- Recursively play the game until maximum value is reached
        go :: LList -> Int -> Int -> M.IntMap Int -> Int
        go node currentMarble currentPlayer scores
          | currentMarble == lastMarble = L.maximum $ M.elems scores
          | (mod currentMarble 23) == 0 =
              let counterClockwiseShifted = counterClockwise7 node
                  updatedScores = (
                    M.adjust
                        (\v -> v + focus counterClockwiseShifted + currentMarble)
                        currentPlayer
                        scores)
                in go (remove counterClockwiseShifted) nextMarble nextPlayer updatedScores
          | otherwise =
                go (insert (clockwise node) currentMarble) nextMarble nextPlayer scores
            where
                nextPlayer = mod (currentPlayer + 1) players
                nextMarble = currentMarble + 1

main :: IO ()
main = print $ solve 435 7118400
-- main = print $ solve 435 71184
